src/coreclr/jit/optimizer.cpp

```diff
@@ -7182,17 +7283,32 @@ void Compiler::optHoistLoopBlocks(unsigned loopNum, ArrayStack<BasicBlock*>* blo
                     {
                         assert(value.Node() != tree);
 
+                        if (!isCommaTree || IsHoistingOverExcepSibling(value.Node(), hasExcep))
```

Comment:
@kunalspathak could you show an example of how this safety check works? It is not obvious from the code why we treat `COMMA`s here specially.

Reply:
The `isCommaTree` check is not really needed since we only set `hasExcep` if `isCommaTree == true`. I have also renamed the method to `IsHoistableOverExcepSibling` to make it more clear.

The idea is to not move `[000332]` before `[000328]` because if it is out of bounds, we would see the exception throw earlier or AV as in https://github.com/dotnet/runtime/pull/68061#issuecomment-1104572027.


```
               [000338] ---XG------                         *  COMMA     void  
     ( 12, 15) [000327] ---XG--H---                         +--*  COMMA     ref    <l:$258, c:$257>
     (  8, 11) [000328] ---X---H---                         |  +--*  BOUNDS_CHECK_Rng void   $245
     (  1,  1) [000329] -------H---                         |  |  +--*  CNS_INT   int    0 $c0
     (  3,  3) [000330] ---X---H---                         |  |  \--*  ARR_LENGTH int    $1c2
     (  1,  1) [000331] -------H---                         |  |     \--*  LCL_VAR   ref    V00 arg0         u:1 $100
     (  4,  4) [000332] n---G--H---                         |  \--*  IND       ref    <l:$256, c:$106>
     (  1,  2) [000333] -------H---                         |     \--*  ARR_ADDR  byref ref[] $82
     (  1,  2) [000334] -------N---                         |        \--*  ADD       byref  $2c0
     (  1,  1) [000335] -------H---                         |           +--*  LCL_VAR   ref    V00 arg0         u:1 $100
     (  1,  1) [000336] -------H---                         |           \--*  CNS_INT   long   16 $280
               [000337] -----------                         \--*  NOP       void  
```

Reply:
Out-of-order hoisting, I see. If we are hosting trees, we must hoist all of the side-effecting ones (in proper execution order), or only hoist the ones without side effects (which I suppose what the logic intends to do).

I still have some questions:
1) Why do we presume this problem can arise for `COMMA`s only?
2) Why is the logic gated on visiting the current tree? The comment above mentions:
```
At this point the stack contains (in top to bottom order):
  ...
  - ancestors of the current node and some of their descendants

  The ancestors have not been visited yet in post order so they're not hoistable
  (and they cannot become hoistable because the current node is not) but some of
  their descendants may have already been traversed and be hoistable.
```
So if some of those "descendants of ancestors" are a mix of such hoistable-not-hoistable nodes (like `[000328]` vs `[000332]` that we have above), the same problem could, presumably, arise.

Finally, as a remark, I have noticed there is quite a devious problem in the tree above, and it is the fact there is not enough information in it to determine that there is interference: `[000332]` _implicitly_ depends on being placed after the range check (it is why it is marked non-faulting). The only thing that saves us there is the fact we have `GTF_GLOB_REF` on it and the interference check is very conservative (`GLOB_REF` by itself doesn't imply any side effects, it just means the tree's value depends on the memory state, but VN already took care of that "in theory"...).

Reply:
> Why do we presume this problem can arise for COMMAs only?

The problem does exist today as well. Today, we don't even check if we are hoisting past a tree that has side-effects (the example that you are stating). To me, `GT_COMMA` was the obvious one because of bounds check, but certainly if there are more examples where things can go wrong, we should address them here or in future PRs. 

> Why is the logic gated on visiting the current tree? 

The same reason. We visit the stack in execution order (bottom to top, as mentioned in comments) and I wanted to make sure to not hoist children past its parents, if parents have side-effects. Let me know if you can think of a better way to validate this.

> the same problem could, presumably, arise.

I noticed last week too while (in `main`) debugging something related where we hoist the checks ahead of side-effects but in that example, the range checks were getting eliminated by assertion prop and hence wouldn't have seen any functional difference.

Reply:
Hmm, so these are essentially "quirks" ("known holes in the algorithm"). They need to be documented as such.

> Let me know if you can think of a better way to validate this.

I would expect the "general" (conservative) logic to be something like:
```cs
bool canHoistExceptions = true; // We can hoist them as long as we haven't left any behind.

for (tree : stack)
{
    bool hoisted = false;

    if (tree->IsHoistable() && (canHoistExceptions || ((tree->gtFlags & GTF_ALL_EFFECT) == 0)))
    {
        hoisted = optHoistCandidate(...); // Can fail for budget reasons, etc
    }

    if (!hoisted && ((tree->gtFlags & GTF_EXCEPT) != 0))
    {
        canHoistExceptions = false;
    }
}
```
Not sure into how many regressions that would run into.

Reply:
> Not sure how many regressions that would run into.

There are lot and it touches lot of methods that contains single-loops too that this PR don't intend to touch and I am not even sure if there are functional issues in these. I will leave it for future PRs.

Benchmarks windows/x64:
`363 total methods with Code Size differences (295 improved, 68 regressed), 58 unchanged.`

Reply:
Yeah, cannot say I didn't suspect something like that would turn up :). My only remaining request here is for comments then.